common/src/app/common/test_helpers/compositions.cljc

```diff
@@ -166,4 +171,143 @@
\ No newline at end of file
+      (thc/instantiate-component component2-label copy2-root-label copy2-root-params)))
+
+(defn- bottom-shape-by-id [file id & {:keys [page-label]}]
+  ;; Get the deepest descendant of a shape by id
+  (let [shape (ths/get-shape-by-id file id :page-label page-label)]
+    (if (some? (:shapes shape))
+      (let [child-id (-> (:shapes shape)
+                         first)]
+        (bottom-shape-by-id file child-id :page-label page-label))
+      shape)))
+
+(defn- bottom-shape  [file tag & {:keys [page-label]}]
+  ;; Get the deepest descendant of a shape by tag
+  (let [shape (ths/get-shape file tag :page-label page-label)]
+    (bottom-shape-by-id file (:id shape) :page-label page-label)))
+
+(defn bottom-fill-color  [file tag & {:keys [page-label]}]
+  ;; Get the first fill color of the deepest descendant of a shape by tag
+  (let [shape (ths/get-shape file tag :page-label page-label)]
+    (-> (bottom-shape file tag :page-label page-label)
+        :fills
+        first
+        :fill-color)))
+
+(defn propagate-component-changes [file component-tag & {:keys [page-label]}]
+  ;; Propagates the component changes for component specified by component-tag
+  (let [page    (if page-label
+                  (thf/get-page file page-label)
+                  (thf/current-page file))
+        file-id (:id file)
+
+        changes (-> (pcb/empty-changes)
+                    (cll/generate-sync-file-changes
+                     nil
+                     :components
+                     file-id
+                     (:id (thc/get-component  file component-tag))
+                     file-id
+                     {file-id file}
+                     file-id))]
+    (thf/apply-changes file changes)))
+
+(defn swap-component [file shape component-tag & {:keys [page-label propagate-fn]}]
+  ;; Swap the specified shape by the component specified by component-tag
+  (let [page    (if page-label
+                  (thf/get-page file page-label)
+                  (thf/current-page file))
+        file-id (:id file)
+        page-id (:id page)
+
+        [new-shape _all-parents changes]
+        (cll/generate-component-swap (pcb/empty-changes)
+                                     (:objects page)
+                                     shape
+                                     (:data file)
+                                     page
+                                     {(:id  file) file}
+                                     (->  (thc/get-component file component-tag)
+                                          :id)
+                                     0
+                                     nil
+                                     {})
+
+        file' (thf/apply-changes file changes)]
+    (if propagate-fn
+      (propagate-fn file')
+      file')))
+
+(defn swap-component-in-shape [file shape-tag component-tag & {:keys [page-label propagate-fn]}]
```

Comment:
I don't like very much having a function that does two different things (this is against Single Responsibility Principle). I'd prefer a function that simply gets the shape, and that can be chained with the other function that does the swap. 

But i haven't found any easy way of creating this chain in the test, that is readable enough.